File: 1.two-sum.ts
/*
 * @lc app=leetcode id=1 lang=typescript
 *
 * [1] Two Sum
 */

// @lc code=start
function twoSum(nums: number[], target: number): number[] {
  const map = new Map<number, number>();
  for (let index = 0; index < nums.length; index++) {
    if (map.get(nums[index]) !== undefined) {
      return [map.get(nums[index])!, index];
    }

    map.set(target - nums[index], index);
  }

  return [];
}
// @lc code=end

function BetterTwoSum(nums: number[], target: number): number[] {
  const numMap = new Map();

  for (let i = 0; i < nums.length; i++) {
    const num = nums[i];
    const difference = target - num;

    if (numMap.has(difference)) return [numMap.get(difference), i];
    else numMap.set(num, i);
  }

  return [];
}

File: 1.two-sum.cpp
/*
 * @lc app=leetcode id=1 lang=cpp
 * @lcpr version=30204
 *
 * [1] Two Sum
 */


// @lcpr-template-start
using namespace std;
#include <algorithm>
#include <array>
#include <bitset>
#include <climits>
#include <deque>
#include <functional>
#include <iostream>
#include <list>
#include <queue>
#include <stack>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
// @lcpr-template-end
// @lc code=start
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> options;

        for (int i = 0; i < nums.size(); i++) {
            if (options.find(target - nums[i]) == options.end()) {
                options[nums[i]] = i;
            } else {
                return {options[target - nums[i]], i};
            }
        }

        return {};
    }
};
// @lc code=end



/*
// @lcpr case=start
// [2,7,11,15]\n9\n
// @lcpr case=end

// @lcpr case=start
// [3,2,4]\n6\n
// @lcpr case=end

// @lcpr case=start
// [3,3]\n6\n
// @lcpr case=end

 */

